File: leet-code/328. Odd Even Linked List/OddEvenLinkedList328.java
public class OddEvenLinkedList328 {
    public ListNode oddEvenList(ListNode head) {
        if (head == null) {
            return head;
        }
        ListNode[] nodes = separate(head);
        while (nodes[0].next != null) {
            nodes[0] = nodes[0].next;
        }
        nodes[0].next = nodes[1];
        return head;
    }

    private ListNode[] separate(ListNode node) {
        if (node == null || node.next == null) {
            return new ListNode[]{node, null};
        }

        ListNode[] nodes = separate(node.next.next);
        ListNode next = node.next;
        node.next = nodes[0];
        next.next = nodes[1];
        return new ListNode[]{node, next};
    }
}

File: leet-code/328. Odd Even Linked List/OddEvenLinkedList328.kt
package com.example

class OddEvenLinkedList328 {
    fun oddEvenList(head: ListNode?): ListNode? {
        if (head == null || head.next == null) {
            return head
        }
        var odd = head
        var even = head.next
        val evenRef = even
        while (even != null && even.next != null) {
            odd!!.next = even.next
            even.next = even.next.next
            // 1->2->3->4
            // 1->2->3

            odd = odd.next
            even = even.next
        }
        odd.next = evenRef
        return head
    }
}
